test/types/for.civet

```diff
@@ -5,47 +5,87 @@ describe "[TS] for", ->
     explicit declaration
     ---
     for var x: T of y
+      console.log x
     ---
-    for (var x: T of y);
+    for (const key of y) {var x: T = key;
+      console.log(x)
+    }
   """
 
   testCase """
     for..of
     ---
     for x: T of y
+      console.log x
     ---
-    for (const x: T of y);
+    for (const key of y) {const x: T = key;
+      console.log(x)
+    }
+  """
+
+  testCase """
+    for each..of
+    ---
+    for each x: T of y
+      console.log x
+    ---
+    for (let i = 0, len = y.length; i < len; i++) {const x: T = y[i];
+      console.log(x)
+    }
   """
 
   testCase """
     for..in
     ---
     for x: T in y
+      console.log x
+    ---
+    for (const key in y) {const x: T = key;
+      console.log(x)
+    }
+  """
+
+  testCase """
+    for..in with two declarations
+    ---
+    for x: T, y in z
+      console.log x, y
     ---
-    for (const x: T in y);
+    for (const key in z) {const x: T = key;const y = z[key];
+      console.log(x, y)
+    }
   """
 
   testCase """
     parenthesized
     ---
     for (x: T of y)
+      console.log x
     ---
-    for (const x: T of y);
+    for (const key of y) {const x: T = key;
+      console.log(x)
+    }
   """
 
   describe ":: typing of properties", ->
     testCase """
       array
       ---
       for [type:: TokenType, token] of Object.entries(tokens)
+        console.log type, token
       ---
-      for (const [type, token]: [ TokenType,unknown] of Object.entries(tokens));
+      for (const key of Object.entries(tokens)) {const [type, token]: [ TokenType,unknown] = key;
```

Comment:
Having the `unknown` here may obstruct TS's inference but I don't know a better way off the top of my head. This is fine for now but we can revisit if we find a better way.

Reply:
`(typeof key)[1]` should work, right?

Reply:
That would work in this instance, but we're currently using a generic "`::` property types to `:` full type" conversion process. I don't think we generally know what we're assigning (e.g. function parameter), though we clearly do in some cases, and maybe we could adapt the derived type in such cases. I'm going to leave this for future work, though.